Mount Point Plugins Implementation

dsheets opened the pull request:
**- What I did**

Designed and implemented mount point interposition plugins as (mostly) described by #33048. This PR implements that proposal but improves the mount point filtering system to be more general and accurate. In particular, filters loaded from mount point plugins are now DNF and include negation (necessary to select local volume bind mounts and anonymous local volume mounts).

I'm moderately happy with how the implementation turned out.

I'm not sure what the Windows story is.

I'm not sure how I should adjust the tests to account for the new API-tests-only policy. Are there good examples somewhere that I should follow? Or a transition PR that I can crib from?

This is my first major contribution to the Docker engine and the most golang I have written in a single patch. I'm very happy to take pointers on both Docker code base best practice and golang best practice.

Container error conditions, daemon restart/reload, and plugin server assumptions were some of my biggest hurdles.

In terms of design, one of my remaining reservations is the level of expressivity of the mount point plugin filter 'language'. It supports what is mostly an equality predicate (set membership for values) and negation. I can see the case for more expressive predicates like path prefix (useful for `--opt device=/prefix/rest`) but this would require expanding the predicate language further (currently negation is a barely acceptable special case).

**- How I did it**

I read a lot of the volume/mount/bind related code and then copied a lot of the structure of the AuthZ plugin subsystem and hooked into the places that seemed relevant from reading and testing.

**- How to verify it**

I wrote some tests in bd35f31. I plan to write some more to test the plugin V2 dynamics (similar to `integration-cli/docker_cli_authz_plugin_v2_test.go`) but the relevant client package isn't published yet (it's in this PR) and I haven't worked out how to fake-vendor it, yet.

I also recommend reading 9d58803 to understand the design.

**- Description for the changelog**

Mount point plugins allow bind mounts and volume mounts to be selectively intercepted

**- A picture of a cute animal (not mandatory but encouraged)**

![marine otter](http://otlibrary.com/wp-content/gallery/marine-otter/080505-003..jpg)

[*Lontra felina*](https://en.wikipedia.org/wiki/Marine_otter), or the marine otter, is believed to be the smallest marine mammal on Earth. It is endangered and can be found along the western coast of South America.

/cc @cpuguy83 @thaJeztah @yallop @dnephin

Commit:
volume: volume.MountPoint.Setup now ensures .Source is set after

Previously, volume.MountPoint.Setup would return the path at which the
mount would appear but the volume.MountPoint object itself might not
contain this path (e.g. in the case of anonymous volume mounts). Now,
.Setup does not return the path and, instead, guarantees the
postcondition that, after a successful call, .Source will contain the
path to the mount point.

Signed-off-by: David Sheets <dsheets@docker.com>

Commit:
volume: fix typo

Signed-off-by: David Sheets <dsheets@docker.com>

Commit:
daemon/volumes: move []container.Mount sorting into container pkg

Signed-off-by: David Sheets <dsheets@docker.com>

Commit:
integration: add mount point plugin tests

Signed-off-by: David Sheets <dsheets@docker.com>

Commit:
volume/mountpoint: implement mount point plugins and middleware

Mount point plugins are just one type of mount point middleware,
now. This will allow us to create internal mount point middleware that
performs similar functions to mount point plugins without all of the
plugin machinery including IPC.

daemon/volumes: split volume.MountPoint.Setup into Realize and Setup

Realize ensures that there is a path to the mount point file system
whereas Setup performs finalization of a mount point including creation
of the mount point path if it does not exist.

volume: moved source path existence validation to mount failure

This simplifies mount spec validation by removing the file system state
input side effect of checking whether the source path exists. Instead,
this error condition is naturally handled during bind mounting. To
achieve this, I introduce a new flag on volume.MountPoint,
CreateSourceIfMissing, that indicates that the source path should be
created if it is missing. We use this flag to decide whether to
idtools.MkdirAllAndChownNew rather than always executing that function
and relying on the early spec validation to prevent --mount requests
from creating directory hierarchies. I also simplified
volume.MountPoint.Setup to remove the optional checking function as this
check can now be performed accurately immediately after
volume.MountPoint.Realize for Unix-like systems.

This change was necessary in order to support mount point plugins that
implement full file system interposition as the requested source should
not be checked early in that case and, instead, the effective source
should be used for mounting (or fail).

daemon/cluster/executor/container: don't precheck bind mount sources

Checking bind mount source directories for existence interferes with
mount point plugins and is unnecessary as automatic creation of source
directories only occurs with 'raw' mount specifications (e.g. `-v
/foo:/bar`) and not with structured specifications.

daemon/cluster/executor/container: add mount point plugins to info

Also, fix an authz plugin type name reporting bug.

volume/mountpoint: add emergency detach instructions for middleware

In the case where middleware is unreachable when it is time to detach an
interposed mount, emergency detach instructions can be used to prevent
mount leaks or file system debris from accumulating or causing further
issues. Without emergency detach instructions, many mount point
middleware chains could fail to detach if a plugin is unreachable (even
if later plugins plowed ahead, they may be unable to clean up their
mounts or directories if the unresponsive plugin used them). With
emergency detach instructions, many mount point plugin chains can safely
and completely detach without _requiring_ responsive plugins. For
resource tracking purposes, we still prefer to contact plugins for
detach.

Signed-off-by: David Sheets <dsheets@docker.com>

Commit:
Mount point plugin docs

Signed-off-by: David Sheets <dsheets@docker.com>

AkihiroSuda commented:
Does it replace volume plugin?

dsheets commented:
@AkihiroSuda this does not replace volume plugins -- it's a different layer from volume plugins. Volume plugins provide some volume file system at a source location. Mount point plugins can then act on this data (or change the source of the mount) before a container starts and after a container stops (but before a return code/exit event is issued). The current implementation allows for a stack of mount point plugins to participate in host bind mounts and volume mounts. The design should be able to accommodate interposing on tmpfs, layers, config, network, and secrets mounts as well (but none of these are implemented yet). I recommend reading #33048 for the motivation and background of this feature.

dsheets commented:
I'm not sure what to do about:

```
These files import internal code: (either directly or indirectly)
15:48:03  - pkg/mountpoint/api.go imports github.com/docker/docker/api/types/mount
15:48:03  - pkg/mountpoint/chain.go imports github.com/docker/docker/api/types/mount
15:48:03  - pkg/mountpoint/chain.go imports github.com/docker/docker/volume
15:48:03  - pkg/mountpoint/mountpoint.go imports github.com/docker/docker/api/types/mount
15:48:03  - pkg/mountpoint/mountpoint.go imports github.com/docker/docker/volume
15:48:03  - pkg/mountpoint/plugin.go imports github.com/docker/docker/api/types/mount
```

Should `pkg/mountpoint` live somewhere else? It is intended to be used by plugins.

Perhaps this highlights an issue with sharing the `volume.MountPoint` serialization via the mount point plugin API because it is not meant to be externally visible?

Should `docker/docker/api` be in a package?

dsheets commented:
I've rebased this PR and moved the implementation from `pkg/mountpoint` into `volume/`. In the process, I've moved some types from `api/types/mount/mount.go` into `pkg/mountpoint/api.go`. This passes the basic validation stages of the CI. The remaining CI failures are cross-platform and easily fixed, spurious, unknown, or intermittent and related. I'm still tracking down the intermittent issue but the rest is ready for design review, afaik.

dnephin reviewed (commented):
Why did the types move to `pkg/mountpoint`? I think these types should be under `api/types`.

Was this because of import checks when you were using `pkg/`? Now that you've moved the logic to `volume/`  I think you can use `api/types` again?

dsheets commented:
The types moved to `pkg/mountpoint` because they are used in the `pkg/mountpoint` API. The types could be duplicated (😢) or `api/types` could be put into a `pkg/`.

dnephin commented:
Why does `pkg/mountpoint` need to be there? How about moving it to `volumes/mountpoint` then you can import from `api/types` ?

dsheets commented:
`pkg/mountpoint` specifies the API for both the client-side and server-side of the mount point plugin API. My understanding was that packages meant to be used by third parties or containing reusable components should live in `pkg/`. Is that not right? I'm just trying to follow the example of `pkg/authorization`...

dnephin commented:
`api/types` is the package for "interface" types that need to be used by both client and server.

In #32989 the proposed new location for `pkg/authorization` is `plugins/authorization`.

I think you're right about the intent of `pkg/` however in practice I don't think that really works. The only guarantee we can make about `pkg/` is that it doesn't import other parts of `docker/docker`.

dsheets commented:
OK, I've put all the new modules under `volume/` and `volume/mountpoint/` and rebased again.

thaJeztah commented on docs/extend/plugins_mountpoint.md:
````diff
@@ -0,0 +1,214 @@
+```json
+{
+    "Success": "Indicates whether the properties query was successful (bool)",
+    "Types":   "Enumerates the types of mount points ('bind', 'volume') this mount point plugin interposes ({ string -> bool })"
````
missing trailing comma (hence the "red" highlighting because it's invalid JSON)

thaJeztah commented on docs/extend/plugins_mountpoint.md:
```diff
@@ -0,0 +1,214 @@
+            with a '!' to invert the match or '\!' to match a key or
+            value beginning with '!'. ({ string -> string })"
+        }
+    ]
```
missing trailing comma

thaJeztah commented on docs/extend/plugins_mountpoint.md:
```diff
@@ -0,0 +1,214 @@
+           "NewMountPoint": "If not omitted or the empty string,
+           indicates a new path to use as the source of this mount point
+           (string)"
+   ],
```
missing closing `}`

thaJeztah commented on docs/extend/plugins_mountpoint.md:
```diff
@@ -0,0 +1,214 @@
+   fatal to the container mount detachment process (false, default) or
+   if the failure should cause the detaching container to return an
+   error code but otherwise continue unwinding the container's mount
+   point plugin stack (bool)"
```
missing trailing comma

thaJeztah reviewed (changes requested):
just some nits found while reading

thaJeztah commented:
ping @johnstep to have a look from the windows side of things

dsheets commented:
I've been prototyping with this PR and I think that the mount point filtering should be much more expressive (e.g. to allow filtering on volume labels) and the attach response message should be more uniform. In particular, I think both should use datatypes with the same shape as the attach request mount point struct.

Regardless of those changes, I believe the basic design of this changeset can be reviewed/approved. I will be off the grid from June 15 until July 7 but @yallop will be available to answer questions.

cpuguy83 reviewed (changes requested):
In general I'm good with design, though I would like to see the pieces split up a bit such that we have mount middleware, and then plugins that can act as mount middleware rather than just mount plugins.
Likewise the plugin API should reflect that the plugins are implementing mountpoint middleware rather than a mount plugin.

dsheets commented on docs/extend/plugins_mountpoint.md:
```diff
@@ -0,0 +1,214 @@
+    "VolumePatterns": [
+        {
+            "VolumePlugin":  "A volume plugin name to require (string)",
+            "OptionPattern": "All of the keys of the OptionPattern must
```
Yes, the filters need to be more expressive than they currently are and packing expressions into strings is not really appropriate (was it ever?). I'm working on this now.

dsheets commented on volume/mountpoint/api.go:
```diff
@@ -0,0 +1,154 @@
+// Type constants
+const (
+	// TypeBind is the type for mounting host dir
+	TypeBind Type = "bind"
```
We could but eventually this enumeration will be bigger than the one in `api/types/mount`. For instance, layers and secrets could also be interposed on. This implementation currently doesn't support `tmpfs` so perhaps that variant should be removed from this type.

stevvooe replied:
This doesn't seem like a very good reason to create yet another type...

stevvooe commented on volume/mountpoint.go:
```diff
@@ -0,0 +1,74 @@
+	return nil
+}
+
+// TopClock returns the Clock value from the plugin on the top of the
```
I'm not sure if I understand the use of the clock terminology here. Is this an ordinal describing the order in which the mounts were applied?

dsheets replied:
Yes, the field doc is `Clock is a positive integer used to ensure mount detachments occur in the correct order`. Because interposition is targeted based on mount point properties, different mounts into the same container can have different stacks of participating plugins. To unwind the plugins in the correct order without maintaining a separate index, we use this ordinal. See https://github.com/moby/moby/pull/33375/files#diff-077bf4d53f2bf30ff32bdb1c2c2c70daR208 for the unwinding routine.

stevvooe replied:
@dsheets It might be good to have some commentary on why this works, over using the separate index approach or just a stack. I don't think I quite understand the intended behavior here.

cpuguy83 replied:
At a glance, this does seem more complicated than other methods. I'm not understanding the unwind code very well.

cpuguy83 commented:
What happens if/when a container exits while the daemon is offline (or is kill 9'd while unwinding or something else equally gruesome)?

Commit:
plugin/store: TEMPORARY DO NOT MERGE add PluginStore.HandleV2

Necessary until #33658 is merged.

Signed-off-by: David Sheets <dsheets@docker.com>

Commit:
daemon/cluster: TEMPORARY DO NOT MERGE add support for mount consistency

This requires an update to swarmkit so the swarm protobuf protocol includes
the consistency field.

Signed-off-by: David Sheets <dsheets@docker.com>

Commit:
integration-cli/docker_cli_plugins_test: test setting a missing mount source

This causes a nil dereference.

Signed-off-by: David Sheets <dsheets@docker.com>

Commit:
plugin/v2: actually set the mount source and device path settings values

Previously, these values which are pointers to strings would be
dereferenced to have their (aliased) strings written to. If these fields
were not specified in the plugin configuration, the values would be
pointers and dereferencing would fail. This removes the use of the
aliased string pointer field and eliminates any mutation side effects.

Signed-off-by: David Sheets <dsheets@docker.com>

Commit:
plugin/v2: settings and not config is the correct source of mount info

Signed-off-by: David Sheets <dsheets@docker.com>

Commit:
plugin/manager: saving a plugin should also update the current plugin store

Signed-off-by: David Sheets <dsheets@docker.com>

Commit:
integration-cli/docker_api_containers_test: fix err is not nil panic

Signed-off-by: David Sheets <dsheets@docker.com>

Commit:
integration-cli/docker_api_containers_test: missing bind source is ok

A missing bind source on container creation is OK. The failure for a
missing bind source should only occur much later after any mount point
middleware has participated and then only if the the mount point has its
CreateSourceIfMissing flag set (e.g. from -v).

Signed-off-by: David Sheets <dsheets@docker.com>

aluzzardi commented:
@dsheets Would it be fair to say, to summarize, that mount point plugins will allow us to:
1) Apply some logic prior to mounting/umounting
2) Accept/Deny mounts

For the latter: isn't that purpose already served by authz plugins?
For the former: Just throwing it here, but wouldn't it make sense to generalize that and just have before/after hooks for whole containers rather than just mount points?

dsheets commented:
@aluzzardi A more general solution is always possible. This PR adds immediately useful functionality built specifically for a similar and pressing set of use cases. A general purpose hook mechanism would need to support a variety of features to accommodate these use cases *and* the use cases described by #14542, #28837, #17021, and #6982. Synchrony, error conditions, startup/shutdown stage, IPC efficiency (both for a given plugin's behavior and for a set of plugins), and failure behavior might all differ for two "general" hooks. For each of these properties, I have designed specific functionality for the mount resource use case. Nothing here precludes the possibility of future generalization.

Authz plugins operate at the API level and have no filters causing IPC for every request made to the daemon, even if not relevant. Additionally, more IPC would be required for a putative authz-based mount point plugin to query the engine for information about volumes or images involved in a mount.

I'm currently porting the test suite from `integration-cli/` to `integration/` and writing documentation. I have multiple plugins written and complete end-to-end implementation in a branch of Docker for Mac. #33658 and #34692 are prerequisites to this constellation of work. I am keen to see this merged in time for 17.10.

dsheets commented:
@cpuguy83 I've renamed the functionality to middleware and made plugins a special case. I've added significant additional functionality and it will be explained in some draft documentation I am preparing.

If the plugin container disappears, the daemon restarts it. The plugin is responsible for its mount state. If the plugin container stops responding when a mount point is detached, `EmergencyDetach` actions are performed which may be sufficient to clean up any leaked resources.

I have a couple of additional error conditions that I'd like to explore including the container death during daemon restart/absence that you mentioned. My current understanding is that the daemon restart logic already handles this case as it must clean up after missing containers when it comes back online. I will double-check that it functions as expected, though.

dnephin commented on integration/util/request/client.go:
```diff
@@ -13,3 +19,115 @@ func NewAPIClient(t *testing.T) client.APIClient {
 }
+
+// Create creates a container with the supplied parameters
+func Create(client client.APIClient, v []string, memoryMB int, labels map[string]string, img string, cmd []string) (string, error) {
```
I don't think the helpers added to this package should be exported from a package. Please move these to the `integration/plugin` package. 

We already have an abstraction for making API calls (the client), any helpers are better off being suite specific instead of trying to maintain another abstraction.

dnephin commented on integration/plugin/mountpoint/mountpoint_unix_test.go:
```diff
@@ -0,0 +1,888 @@
+}
+
+func daemonIsLinux(t *testing.T) {
+	if testEnv.DaemonInfo.OSType != "linux" {
```
Instead of hiding the checks in these helpers I think it's more obvious to use `gotestyourself/skip`:

(in the test case):
```go
skip.IfCondition(t, testEnv.DaemonInfo.OSType != "linux")
```

same for the two above.

dnephin commented on integration/plugin/mountpoint/mountpoint_unix_test.go:
```diff
@@ -0,0 +1,888 @@
+
+			//logs, err := d.ReadLogFile()
+			//require.Nil(c, err)
+			//fmt.Print(string(logs))
```
commented out code

dnephin commented on integration/plugin/mountpoint/mountpoint_unix_test.go:
```diff
@@ -723,19 +719,19 @@ func TestMountPointPluginKill(c *testing.T) {
-	swapMemorySupport(c)
+	skip.IfCondition(c, testEnv.DaemonInfo.OSType != "linux")
+	skip.IfCondition(c, !sysInfo.MemoryLimit)
+	skip.IfCondition(c, !sysInfo.SwapLimit)
```
This is checking `sysInfo` for the host where the tests are running, but the docker engine may be running on a different host. Does this test need to check `testEnv.IsLocalDaemon()` as well?

dsheets replied:
I cribbed this from <https://github.com/moby/moby/blob/master/integration-cli/docker_cli_oom_killed_test.go#L11>. I'm not aware of additional requirements for the tests to support non local daemons. I can add the check if you'd like.

thaJeztah commented:
Tests are filing

dsheets commented:
@thaJeztah I don't understand why they are failing due to busybox. They pass on my machine. :-/

dnephin commented on integration/plugin/mountpoint/mountpoint_unix_test.go:
```diff
@@ -0,0 +1,866 @@
+}
+
+func setupTest(c *testing.T) func() {
+	testEnv.ProtectImage(c, "busybox:latest")
```
This is why the tests are failing. This needs to be:

```go
environment.ProtectImages(t, testEnv)
```

dnephin commented on integration/plugin/mountpoint/mountpoint_unix_test.go:
```diff
@@ -0,0 +1,866 @@
+			{Type: &typeBind},
+		},
+	}
+	// matches -v /host:/container AND all local volume mounts
```
Could you split this out into functions with descriptive names instead of these inline comments?

Then `ctrl` (also needs a more descriptive name) can be created by calling the factory functions.

dnephin replied:
Even better, if all of these aren't used for every test could they be created in the test that uses them?

dsheets commented:
It looks like a defect was introduced in the rebase on top of #34258 that causes automatic creation of source directories to fail. I'll take a look tomorrow.

dnephin commented on integration/plugin/mountpoint/main_test.go:
```diff
@@ -0,0 +1,33 @@
+	sysInfo = sysinfo.New(true)
+
+	testEnv.Print()
+	setupSuite()
```
This is going to break windows CI (I just realized that these suites aren't running on windows CI but that should be fixed shortly).

It looks like `mountpoint_windows_test.go` is missing which needs to define these.

dnephin commented on integration/plugin/mountpoint/mountpoint_unix_test.go:
```diff
@@ -0,0 +1,899 @@
+	mountPointController[2] = mountPointControllerForLocalVolumeBindMounts()
+	mountPointController[3] = mountPointControllerForAnonymousBindMounts()
+	mountPointController[4] = mountPointControllerForAllBindMounts()
+	mountPointController[5] = mountPointControllerForMountsIntoContainersWithoutAllowLabel()
```
It seems that there are 6 controllers being setup, but some tests also modify the controllers. I believe not every test uses every controller, is that right?

I think this is going to be difficult to maintain. Instead of having everything setup for every test could you have a function like:

```go
func setupPlugin(controllers ...testMountPointController) { ... }
```

and call that on each test with just the controllers that it needs? That should also remove the need to modify the static controllers from individual tests, and will hopefully remove the indexing using numbers as well.

`setupPlugin()` can still add them to a static slice, and teardown can iterate over the slice to perform cleanup.

dsheets commented:
Could someone please review the design and implementation of this PR?

A few additional test cases and some minor clean-up will be forthcoming but nothing is blocking a review, afaik.

flx42 commented:
FWIW, we (NVIDIA, for GPU enablement in containers) would not benefit from this new type of plugin, I believe. Having support for standardized OCI hooks would be better for us. As described in #28837, such hooks could either be registered to the docker daemon (like we do for runtimes), or passed to `docker run` with something like `docker run --exec-opt prestart=hook1,hook2`, or dropped inside a directory like in Red Hat's fork.

cpuguy83 commented:
@flx42 I think hooks are basically deprecated in favor of the create/start split that is available in runc.... how to expose that from docker I'm really not certain.... right now there isn't even such a split in the moby codebase yet.

flx42 commented:
@cpuguy83 What do you mean "deprecated"? Aren't they part of the OCI runtime spec?
https://github.com/opencontainers/runtime-spec/blob/v1.0.1/specs-go/config.go#L122-L130

cpuguy83 commented:
@flx42 Sure it's part of the spec, but it basically the preference is to use create/start instead of hooks. Deprecated doesn't mean gone.

tiborvass commented:
@anusha-ragunathan it would be interesting to investigate whether this approach could work for @flx42 as well.

vdemeester commented:
@cpuguy83 what is the status here ?

cpuguy83 commented:
If this is something we need to have, it would be good if someone who is interested in having this merged to take it over.
Also are we sure this is the right layer to abstract? Could (or should) this be generalized for the entire container (i.g. full container lifecycle hooks)?

I'm going to go ahead and close this one (which has tons of merge conflicts as is) and move discussion back to the open issue. Thanks! 👼
